Add Patch resources to all SDKs

lblackstone opened the pull request:
<!--Thanks for your contribution. See [CONTRIBUTING](CONTRIBUTING.md)
    for Pulumi's contribution guidelines.

    Help us merge your changes more quickly by adding more details such
    as labels, milestones, and reviewers.-->

### Proposed changes
Add Patch resources for all Kubernetes resource types that can be used to patch existing cluster resources rather than creating a new resource under Pulumi management. See https://github.com/pulumi/pulumi-kubernetes/issues/2011 for additional details on the background and design of this feature.
<!--Give us a brief description of what you've done and what it solves. -->

### Related issues (optional)
Fix https://github.com/pulumi/pulumi-kubernetes/issues/2015
<!--Refer to related PRs or issues: #1234, or 'Fixes #1234' or 'Closes #1234'.
    Or link to full URLs to issues or pull requests in other GitHub repositories. -->

lblackstone commented:
@viveklak It looks like tests may be failing due to parallel resource creation across SDKs. I'll take another look tomorrow, but they're passing for me locally.

viveklak commented on provider/pkg/await/await.go:
```diff
@@ -181,8 +181,9 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
-					Force:        &force,
+					FieldManager:    c.FieldManager,
+					Force:           &force,
+					FieldValidation: metav1.FieldValidationIgnore,
```
Do we want this to be warn instead?

viveklak commented on tests/sdk/nodejs/server-side-apply/step2/index.ts:
```diff
@@ -0,0 +1,180 @@
+            labels: {
+                foo: "foo",
+            },
+            name: "my-new-cron-object",
```
Just wanted to make sure we are not letting patch resources by autonamed in the provider? I did notice that we are not enforcing its requirement in the SDK generation (which maybe we ought to do?)

lblackstone replied:
Good point. The name should be required.

lblackstone commented on provider/pkg/gen/schema.go:
```diff
@@ -450,6 +531,17 @@ func genPropertySpec(p Property, resourceGV string, resourceKind string) pschema
 	err := json.Unmarshal([]byte(p.SchemaType()), &typ)
 	contract.Assert(err == nil)
 
+	if strings.HasSuffix(resourceKind, "Patch") {
```
Yeah, there's nothing currently, and I think we could handle that case if it did arise in a later release. I don't think it's particularly likely.

viveklak commented on provider/pkg/gen/schema.go:
```diff
@@ -294,6 +298,44 @@ func PulumiSchema(swagger map[string]interface{}) pschema.PackageSpec {
+
+					var propNames []string
+					for _, p := range kind.Properties() {
+						patchSpec.Properties[p.name] = genPropertySpec(p, kind.apiVersion, kind.kind+"Patch")
```
Somewhat subtle that no required properties are being set (by design for patch resources). Might be worth calling out with a comment. We might want to add name and/or namespace (for namespaced objects back though)?

viveklak commented on provider/pkg/provider/provider.go:
```diff
@@ -2683,12 +2707,13 @@ func (k *kubeProvider) tryServerSidePatch(
 	}
-	if err != nil {
-		return nil, nil, false, err
+	if err.Error() == "name is required" {
```
Nit - consider returning a strongly typed error for this instead?

viveklak reviewed (approved):
Directionally LGTM! We might move addressing some of the suggestions to a subsequent PR for review ergonomics.

lblackstone commented on provider/pkg/gen/dotnet-templates/apiextensions/CustomResourcePatch.cs:
```diff
@@ -0,0 +1,133 @@
+using Pulumi.Kubernetes.Types.Inputs.Meta.V1;
+using Pulumi.Kubernetes.Types.Outputs.Meta.V1;
+
+namespace Pulumi.Kubernetes.ApiExtensions
```
I just copy/pasted the implementation from CustomResource and renamed some relevant fields. I think it's ok for an initial pass here.

lblackstone commented:
> Directionally LGTM! We might move addressing some of the suggestions to a subsequent PR for review ergonomics.

Sounds good! I will make another pass through these feedback items once this merges.
